### PR TITLE
[VectorExt] Add layout iterator classes

### PR DESCRIPTION
**Pull request by harsh-nod**
This PR adds iterator classes to iterate over the layout and to concatenate iterators with other frozen iterators. This is required when distributing reductions. Also adds a test to check for correctness.

### CODE REVIEW

File: llvm-external-projects/iree-dialects/lib/Dialect/VectorExt/IR/VectorExtOps.cpp [hunk 1]
```diff
@@ -52,7 +52,84 @@ OpFoldResult ToSIMTOp::fold(FoldAdaptor) {
...
+}
+
+LayoutIterator::LayoutIterator(PerDimLayoutAttr &attr,
+                               DenseMap<LayoutDimension, int64_t> strides) {
+  initialize(attr, strides);
+}
+
+/// The iterator is done when it returns back to
+/// its begin state.
+bool LayoutIterator::iterationComplete() {
```
Comment (Groverkss) [inline llvm-external-projects/iree-dialects/lib/Dialect/VectorExt/IR/VectorExtOps.cpp:111]: nit: iterationComplete -> isIterationComplete

**Review by Groverkss — APPROVED** [review]
LGTM, but I didn't get the difference between a frozen iterator and an unfrozen one.

**Comment by harsh-nod** [comment]
> LGTM, but I didn't get the difference between a frozen iterator and an unfrozen one.

That was to support the "partial iterator". So say you want to iterate over only the parallel dimensions but want to "freeze" the reduction dims at a certain point in the iteration space, then you would use that provided function.

File: llvm-external-projects/iree-dialects/lib/Dialect/VectorExt/IR/VectorExtOps.cpp (hunk 1 above)
Comment (harsh-nod) [reply llvm-external-projects/iree-dialects/lib/Dialect/VectorExt/IR/VectorExtOps.cpp:111]: So I was thinking about how this reads
```
do {
  ...
}while (!iterationComplete());
```
which is _do something while not iteration complete_ which to me reads better than _do something while not is iteration complete_. But can change if you feel strongly.

**Pull request merged**